### DISCUSSION TITLE
JetStream Direct Get Batch / Multi Last

### DISCUSSION THREAD

**Discussion by charbonnierg** [nats-io/nats.py#550] [Show and tell]
Hi everyone,

[NATS 2.11 is coming](https://github.com/nats-io/nats-server/releases/tag/v2.11.0-preview.1) soon, and will bring 2 new features:

- [JetStream Direct Get Batch Requests](https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-31.md#batched-requests)
- [JetStream Direct Get Multi-subject Requests](https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-31.md#multi-subject-requests)

Those two features have been implemented in [jsm.go](https://github.com/nats-io/jsm.go/blob/1c7f0e42497859a4594efbc657ca25237cab7359/streams.go#L749) in [this PR](https://github.com/nats-io/jsm.go/pull/530/files).

I tried to come up with [an implementation in Python](https://github.com/nats-io/nats.py/commit/d7997c05d0c3eaf1b64334a59445dcfc8540f168), here's how it's used: https://github.com/charbonnierg/nats.py/blob/feat/multi_direct_get/examples/direct_get.py

What do you think about this API ?

@wallyqs would you be interested in a PR ? If so, could you please describe the kind of tests that you would expect, and if it would be ok to write those tests with `pytest` in a new module rather than bloating the existing `test_js.py` which is already > 4000 LOC

@bruth I know you're not actively contributing to this repo, but your opinion could also be extremely useful 😊

Also, I guess that once the batch direct get is supported, it should also be possible to implement a `get_many` or  something similar for KV stores, but I don't have a strong opinion on the API to use, neiter did I look into other client implementations to get an idea.

**Comment by wallyqs** [comment]
@charbonnierg a PR here would be very welcome! How about adding the test in a different file bucketing the v2.11 features? like `test_js_v2_11.py`?

**Reply by charbonnierg** [reply]
Sounds like a good idea 👍 I'll try to do that then !
Would you be ok with tests using pytest ? Since pytest is used to discover and run tests I don't think that should be a problem. I find @pytest.mark.parametrize to be extremely useful when writing test cases.

**Reply by wallyqs** [reply]
that sounds good to me, using pytest is fine
